### PR TITLE
Add Dockerfile and github action for docker hub

### LINKED ISSUE #13: Github action to create and push docker container
**Is your feature request related to a problem? Please describe.**
I think that it would be useful to add a GitHub action that automatically builds a docker image and pushes it to docker hub. In this way, we can replace the Dockerfile build [here](https://github.com/zavolanlab/Dockerfiles/tree/master/MultiQC) and create separate tags depending on the tools we have in the current repository. The problem that we have now is that if we update the multiqc plug in here (e.g. with an addition of a new tool) we have to use the same tag in the Dockerfiles repository.

**Describe the solution you'd like**
- [x] Add Dockerfile in the repo
- [x] Add dockerhub credentials in the repo
- [x] Add github action that is triggered when the tag starts with **v**

### PR DESCRIPTION
**Pull request by fgypas**
## Description

Add Dockerfile and GitHub action that automatically builds and pushes an image to Dockerhub after making a release that starts with **v**.

Fixes #13

## Type of change

**Describe the solution you'd like**
- [X] Add Dockerfile in the repo
- [X] Add dockerhub credentials in the repo
- [C] Add github action that is triggered when the tag starts with **v**

### CODE REVIEW

**Comment by fgypas** [comment]
Here is an example of the tag:
https://github.com/zavolanlab/multiqc-plugins/releases/tag/v0.01_test_github_actions (we can delete later)

This means that when I tagged `v0.01_test_github_actions` it automatically pushed the following to docker hub:
https://hub.docker.com/layers/131312112/zavolab/multiqc-plugins/0.01_test_github_actions/images/sha256-967c52fdf7b3e127c23003b1b42a5a483eb2f670368cc7176f8b18bd99f023da?context=explore

The advantage of doing it like this is that we can have tags that are more descriptive.
For example if we tag:
`vmultiqc_1.9_alfa_tin`

we can pull the images with the following:
docker://zavolab/multiqc-plugins:multiqc_1.9_alfa_tin

@AngryMaciek Let me know what you think. Then we can also use this directly to [zarp](https://git.scicore.unibas.ch/zavolan_group/pipelines/zarp) or other projects.

File: .github/workflows/build-image.yml [hunk 1]
```diff
@@ -0,0 +1,26 @@
+name: Docker registry
+
+on: 
+   push:
```
Comment (AngryMaciek) [inline .github/workflows/build-image.yml]: incorrect indentation :P

File: .github/workflows/build-image.yml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+     tags:
+       - 'v*'
+
+jobs:
+  github:
```
Comment (AngryMaciek) [inline .github/workflows/build-image.yml]: better name for the job would be sth. like `push-dockerhub` (than `github`)

**Review by AngryMaciek — COMMENTED** [review]
Looks nice, great idea.

I am a little confused how does it work. Since you trigger this action `on` events: `push | tags | v* ` I presume that it should be called whenever you decide to create a `v*` tagged release out of any commit in any branch? So this GH workflow runs additionally after creating releases via the GitHub interface and will not be a part of a 'standard' commit-and-push-related CI, correct? But if so, why is it set  `on push`... <?>

So I am having some mixed interpretation of `push` and `tag` events...Maybe we should use this one instead?
https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#release

**Comment by AngryMaciek** [comment]
I misunderstood the concept of tagging. I red a little and understand now that we can tag commits locally and then push them to trigger the CI including the Docker publish. I guess this is what you had in mind?

**Comment by fgypas** [comment]
Yes, exactly. It will not run on every commit or push you make. Only when you push a tag, the dockerhub action will be triggered. In this way, you have all the "power" in one repository and you don't depend on others when you update.

File: .github/workflows/build-image.yml (hunk 1 above)
Comment (fgypas) [reply .github/workflows/build-image.yml]: I think it's fixed now.

File: .github/workflows/build-image.yml (hunk 2 above)
Comment (fgypas) [reply .github/workflows/build-image.yml]: I changed it to `push_dockerhub`

**Review by AngryMaciek — COMMENTED** [review]
It's all good in the code.

How does the order of operations in gitflow follow now? From the trigger I see that the workflow will execute upon push of **any** tagged commit (starting with v) to **any** branch...

1) create a feature branch
2) add a feature
3) start a PR
4) PR reviewed as OK (we are here now)
5) Merge?

So where do we put in the 'tagged commit'? I guess it cannot be before (5) since exactly that step accepts the new feature. Tagging a version before that would result in a version available on PyPI before the feature is accepted here... On the other hand - while merging you delete the feature branch. I see two possibilities here:
* push directly to `master` after each PR
* create a new feature branch `ver.xxx` just for the new version, push the tagged commit there and then merge it. This is still 'reverse order' than it should but the feature itself was already accepted in the previous PR.

In both cases the changes to the repo would be **only** a version bump.

Do you have any other ideas?

**Comment by fgypas** [comment]
- So here we will merge to develop. 
- I think then you should merge to master/main
- And then tag the latest commit of master/main 

In general, a good strategy is to use master/main for tag releases.

**Comment by fgypas** [comment]
Sorry, I pressed the wrong button and re-run the CI. It should be fine now.

**Comment by AngryMaciek** [comment]
> * So here we will merge to develop.
> * I think then you should merge to master/main
> * And then tag the latest commit of master/main
> 
> In general, a good strategy is to use master/main for tag releases.

So I understand that after merging an updated `develop` to `master` you propose to `git pull` locally then `git checkout master`, bump up the versions, `git commit`, tag that commit and then `git push origin master --follow-tags` ?

In such solution the automatic relsease from the CI would be triggered by a **direct commit** to the `master`.

**Comment by fgypas** [comment]
@AngryMaciek Yes, this can be done. 

Alternatively, you can bump-up the versions here and merge to develop and then merge to master.
Then you can just create a tag manually from the UI https://github.com/zavolanlab/multiqc-plugins/releases/new

Both methods would work

**Comment by AngryMaciek** [comment]
@fgypas 

>Alternatively, you can bump-up the versions here and merge to develop and then merge to master.
Then you can just create a tag manually from the UI https://github.com/zavolanlab/multiqc-plugins/releases/new

Are you sure this will work?

The CI for _Docker_ build is triggered by:

```
on: 
  push:
    tags:
      - 'v*'
```

Therefore, as manual tagging from the UI would happen **after** all the pushes & merges - it will not trigger a GH Action, no?

**Comment by fgypas** [comment]
@AngryMaciek When you do a tag it will trigger only this Github action: https://github.com/zavolanlab/multiqc-plugins/pull/14/files#diff-8b50c1c8ac07bc956388add66e256e8336ae22fd2cac8d83f770e1dc63674c00

This is because the action starts only when the tag starts with a `v` (see here https://github.com/zavolanlab/multiqc-plugins/blob/feature/13-dockerhub-github-action/.github/workflows/build-image.yml#L6)

We can also discuss quickly over zoom if you prefer

**Comment by AngryMaciek** [comment]
This issue has been clarified on _zoom_ : manual tagging on GitHub also triggers the CI dockehub push.  
Ready to merge.

**Pull request merged**